components/raftstore/src/store/util.rs

```diff
@@ -1123,6 +1125,65 @@ pub fn check_conf_change(
     }
 }
 
+fn check_remove_or_demote_voter(
+    cfg: &Config,
+    change_peers: &[ChangePeerRequest],
+    leader_id: u64,
+    peer_heartbeat: &collections::HashMap<u64, std::time::Instant>,
+) -> Result<()> {
+    // max heartbeats missed to qualify being a slow peer
+    const MAX_MISSED_HEARTBEATS: u32 = 8;
+    let mut slow_peer_count = 0;
+    let mut normal_peer_count = 0;
+    for (id, last_heartbeat) in peer_heartbeat {
+        if *id == leader_id
+            || last_heartbeat.elapsed() <= MAX_MISSED_HEARTBEATS * cfg.raft_heartbeat_interval()
+        {
+            normal_peer_count += 1;
+        } else {
+            slow_peer_count += 1;
+        }
+    }
+
+    let mut normal_peers_to_remove = vec![];
+    for cp in change_peers {
+        let (change_type, peer) = (cp.get_change_type(), cp.get_peer());
+        if change_type == ConfChangeType::RemoveNode
+            || change_type == ConfChangeType::AddLearnerNode
```

Comment:
Could you further check if the demote peer is actually in the region peer list? So it does not accidentally block adding peers.

Reply:
@overvenus Yes, it's checked in the code below. We thought it's a demote only when peer_beartbeat finds the peer_id